### PR TITLE
Fix pickle/deepcopy node hole handling

### PR DESCRIPTION
**Pull request by mtreinish**
This commit fixes an issue introduced by #589 where in certain cases node holes in a graph would result in a panic being raised. This was caused by a logic bug in trying to recreate the holes. Additionally, there were several places where graph methods removed nodes that the flag to indicate there were removals would no be set. This commit fixes all of these issues so that deepcopy/pickle works as expected.

### CODE REVIEW

File: src/digraph.rs [hunk 1]
```diff
@@ -384,7 +383,7 @@ impl PyDiGraph {
             let node_idx: usize = item.get_item(0).unwrap().extract().unwrap();
             let node_w = item.get_item(1).unwrap().extract().unwrap();
 
-            for _i in 0..node_idx {
+            for _i in 0..node_idx + 1 {
```
Comment (mtreinish) [inline src/digraph.rs:386]:
```suggestion
            for _i in 0..node_idx {
```

File: src/graph.rs [hunk 2]
```diff
@@ -284,56 +285,24 @@ impl PyGraph {
...
             let mut tmp_nodes: Vec<NodeIndex> =
                 Vec::with_capacity(node_bound_1 + 1 - nodes_lst.len());
 
-            for i in 0..nodes_lst.len() + 1 {
-                if i < next_node_idx {
+            for item in nodes_lst {
+                let item = item.downcast::<PyTuple>().unwrap();
+                let next_index: usize = item.get_item(0).unwrap().extract().unwrap();
+                let weight: PyObject = item.get_item(1).unwrap().extract().unwrap();
+                while next_index > self.graph.node_bound() {
```
Comment (enavarro51) [inline src/graph.rs:297]: What's the purpose of this change? (This review may be mostly questions and same comments will apply to graph.rs and digraph.rs.)

File: src/graph.rs [hunk 3]
```diff
@@ -1062,8 +1019,8 @@ impl PyGraph {
     ///     the graph
     #[pyo3(text_signature = "(self, index_list, /)")]
     pub fn remove_nodes_from(&mut self, index_list: Vec<usize>) -> PyResult<()> {
-        for node in index_list.iter().map(|x| NodeIndex::new(*x)) {
-            self.graph.remove_node(node);
+        for node in index_list {
+            self.remove_node(node)?;
```
Comment (enavarro51) [inline src/graph.rs:1023]: Why all the changes from `self.graph.remove_node` to `self.remove_node`?

File: src/graph.rs [hunk 4]
```diff
@@ -1846,7 +1803,10 @@ impl PyGraph {
 
     fn __delitem__(&mut self, idx: usize) -> PyResult<()> {
         match self.graph.remove_node(NodeIndex::new(idx)) {
-            Some(_) => Ok(()),
+            Some(_) => {
+                self.node_removed = true;
+                Ok(())
+            }
```
Comment (enavarro51) [inline src/graph.rs:1809]: This seems pretty obvious. Was there some reason this wasn't done previously?

File: src/graph.rs (hunk 3 above)
Comment (mtreinish) [reply src/graph.rs:1023]: This is the second order issue (which the release note is about) where we weren't tracking removals correctly in several methods. Using `self.remove_node` does the tracking for us.

File: src/graph.rs (hunk 4 above)
Comment (mtreinish) [reply src/graph.rs:1809]: It was an oversight that's the second fix around hole handling (and what I added the release note for)

**Review by enavarro51 — APPROVED** [review]
Ok. LGTM. Thanks.
